# Phase 1 — Data Understanding & Profiling

# Setup

Importing libraries

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)
pd.set_option("display.width", 120)

# 2. Table viewing and basic info
Viewing the first 5 rows of the table along with count total of rows and columns

In [4]:
df1 = pd.read_excel("../data/raw/online_retail_II.xlsx", sheet_name="Year 2009-2010")
print(f"year_2009_2010: {df1.shape[0]} rows, {df1.shape[1]} columns")
display(df1.head(5))

year_2009_2010: 525461 rows, 8 columns


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


In [5]:
df2 = pd.read_excel("../data/raw/online_retail_II.xlsx", sheet_name="Year 2010-2011")
print(f"year_2010_2011: {df2.shape[0]} rows, {df2.shape[1]} columns")
display(df2.head(5))

year_2010_2011: 541910 rows, 8 columns


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [8]:
# Inspecting the data types and null values
print(f"table: retail_2009_2010")
df1.info()

table: retail_2009_2010
<class 'pandas.DataFrame'>
RangeIndex: 525461 entries, 0 to 525460
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      525461 non-null  object        
 1   StockCode    525461 non-null  object        
 2   Description  522533 non-null  object        
 3   Quantity     525461 non-null  int64         
 4   InvoiceDate  525461 non-null  datetime64[us]
 5   Price        525461 non-null  float64       
 6   Customer ID  417534 non-null  float64       
 7   Country      525461 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 32.1+ MB


In [9]:
# Inspecting the data types and null values
print(f"table: retail_2010_2011")
df2.info()

table: retail_2010_2011
<class 'pandas.DataFrame'>
RangeIndex: 541910 entries, 0 to 541909
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   Invoice      541910 non-null  object        
 1   StockCode    541910 non-null  object        
 2   Description  540456 non-null  object        
 3   Quantity     541910 non-null  int64         
 4   InvoiceDate  541910 non-null  datetime64[us]
 5   Price        541910 non-null  float64       
 6   Customer ID  406830 non-null  float64       
 7   Country      541910 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


# 3. Appending both the tables

In [12]:
retail_transactions = pd.concat([df1, df2], ignore_index=True)
print(f"retail_transactions: {retail_transactions.shape[0]} rows, {retail_transactions.shape[1]} columns")
display(retail_transactions.head(5))

retail_transactions: 1067371 rows, 8 columns


,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489434,85048,15CM CHRISTMAS GLASS BALL 20 LIGHTS,12,2009-12-01 07:45:00,6.95,13085.0,United Kingdom
1,489434,79323P,PINK CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
2,489434,79323W,WHITE CHERRY LIGHTS,12,2009-12-01 07:45:00,6.75,13085.0,United Kingdom
3,489434,22041,"RECORD FRAME 7"" SINGLE SIZE",48,2009-12-01 07:45:00,2.10,13085.0,United Kingdom
4,489434,21232,STRAWBERRY CERAMIC TRINKET BOX,24,2009-12-01 07:45:00,1.25,13085.0,United Kingdom


# 4. Missing Audit

In [ ]:
# Inspecting null values in each column of the retail_transactions DataFrame
print(f"table: retail_transactions")
print(retail_transactions.isnull().sum())

table: retail_transactions
Invoice             0
StockCode           0
Description      4382
Quantity            0
InvoiceDate         0
Price               0
Customer ID    243007
Country             0
dtype: int64


In [ ]:
# Calculating the percentage of null values in each column of the retail_transactions DataFrame
print(f"table: retail_transactions")
null_values = (retail_transactions.isnull().mean() * 100).round(2)
print(null_values[null_values > 0].sort_values(ascending=False))

table: retail_transactions
Customer ID    22.77
Description     0.41
dtype: float64


In [ ]:
# Identifying rows with null values in both "Description" and "Customer ID" columns
retail_transactions[retail_transactions["Description"].isnull() & retail_transactions["Customer ID"].isnull()].reset_index(drop=True)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
1,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
2,489659,21350,NaN,230,2009-12-01 17:39:00,0.0,NaN,United Kingdom
3,489781,84292,NaN,17,2009-12-02 11:45:00,0.0,NaN,United Kingdom
4,489806,18010,NaN,-770,2009-12-02 12:42:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
4377,581199,84581,NaN,-2,2011-12-07 18:26:00,0.0,NaN,United Kingdom
4378,581203,23406,NaN,15,2011-12-07 18:31:00,0.0,NaN,United Kingdom
4379,581209,21620,NaN,6,2011-12-07 18:35:00,0.0,NaN,United Kingdom
4380,581234,72817,NaN,27,2011-12-08 10:33:00,0.0,NaN,United Kingdom


# 5. Duplicate checks

In [ ]:
print(f"table: retail_transactions")
print(f"duplicates: {retail_transactions.duplicated().sum()}")  # Count of duplicate rows
print(f"{retail_transactions.duplicated().mean() * 100:.2f}% of rows are duplicates")   # Percentage of duplicate rows

table: retail_transactions
duplicates: 34335
3.22% of rows are duplicates


# 6. Categorical value inspection
Checking the distribution of categorical variables

In [58]:
retail_transactions["Country"].value_counts(dropna=False)

Country
United Kingdom          981330
EIRE                     17866
Germany                  17624
France                   14330
Netherlands               5140
Spain                     3811
Switzerland               3189
Belgium                   3123
Portugal                  2620
Australia                 1913
Channel Islands           1664
Italy                     1534
Norway                    1455
Sweden                    1364
Cyprus                    1176
Finland                   1049
Austria                    938
Denmark                    817
Unspecified                756
Greece                     663
Japan                      582
USA                        535
Poland                     535
United Arab Emirates       500
Israel                     371
Hong Kong                  364
Singapore                  346
Malta                      299
Iceland                    253
Canada                     228
Lithuania                  189
RSA                        169


From the categorical value inspection it is found that, we are getting maximum customers from UK, so the data would naturally be more inclined towards UK.

# 7. Numerical value check
Descriptive statistics for numerical columns

In [62]:
retail_transactions.describe(exclude= "object").round(2)

,Quantity,InvoiceDate,Price,Customer ID
count,1067371.00,1067371,1067371.00,824364.00
mean,9.94,2011-01-02 21:13:55.394029,4.65,15324.64
min,-80995.00,2009-12-01 07:45:00,-53594.36,12346.00
25%,1.00,2010-07-09 09:46:00,1.25,13975.00
50%,3.00,2010-12-07 15:28:00,2.10,15255.00
75%,10.00,2011-07-22 10:23:00,4.15,16797.00
max,80995.00,2011-12-09 12:50:00,38970.00,18287.00
std,172.71,NaN,123.55,1697.46


# 8. Profiling of few columns

In [ ]:
# Identifying rows with "Invoice" starting with "C" as they represent canceled transactions
retail_transactions[retail_transactions["Invoice"].str.startswith("C", na=False)].reset_index(drop=True)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,C489449,22087,PAPER BUNTING WHITE LACE,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
1,C489449,85206A,CREAM FELT EASTER EGG BASKET,-6,2009-12-01 10:33:00,1.65,16321.0,Australia
2,C489449,21895,POTTING SHED SOW 'N' GROW SET,-4,2009-12-01 10:33:00,4.25,16321.0,Australia
3,C489449,21896,POTTING SHED TWINE,-6,2009-12-01 10:33:00,2.10,16321.0,Australia
4,C489449,22083,PAPER CHAIN KIT RETRO SPOT,-12,2009-12-01 10:33:00,2.95,16321.0,Australia
...,...,...,...,...,...,...,...,...
19489,C581490,23144,ZINC T-LIGHT HOLDER STARS SMALL,-11,2011-12-09 09:57:00,0.83,14397.0,United Kingdom
19490,C581499,M,Manual,-1,2011-12-09 10:28:00,224.69,15498.0,United Kingdom
19491,C581568,21258,VICTORIAN SEWING BOX LARGE,-5,2011-12-09 11:57:00,10.95,15311.0,United Kingdom
19492,C581569,84978,HANGING HEART JAR T-LIGHT HOLDER,-1,2011-12-09 11:58:00,1.25,17315.0,United Kingdom


In [ ]:
# Identifying rows with "Invoice" starting with "C" and null "Customer ID"
retail_transactions[(retail_transactions["Invoice"].str.startswith("C", na=False)) & 
                    (retail_transactions["Customer ID"].isnull())].reset_index(drop=True)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,C489859,M,Manual,-1,2009-12-02 14:45:00,69.57,NaN,United Kingdom
1,C489860,90029,NEW BAROQUE SMALL NECKLACE BLACK,-1,2009-12-02 14:49:00,8.50,NaN,United Kingdom
2,C489881,21102,"CANDLE IN JAR COLLAGE , VANILLA",-31,2009-12-02 16:20:00,2.55,NaN,Nigeria
3,C489881,72802A,ROSE SCENT CANDLE IN JEWELLED BOX,-16,2009-12-02 16:20:00,4.25,NaN,Nigeria
4,C490307,POST,POSTAGE,-1,2009-12-04 14:49:00,8.97,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
745,C578097,22112,CHOCOLATE HOT WATER BOTTLE,-48,2011-11-22 17:31:00,4.25,NaN,United Kingdom
746,C579757,47469,ASSORTED SHAPES PHOTO CLIP SILVER,-24,2011-11-30 14:56:00,0.65,NaN,United Kingdom
747,C579907,22169,FAMILY ALBUM WHITE PICTURE FRAME,-2,2011-12-01 08:48:00,7.65,NaN,EIRE
748,C580604,AMAZONFEE,AMAZON FEE,-1,2011-12-05 11:35:00,11586.50,NaN,United Kingdom


In [ ]:
# Identifying rows with "Invoice" not starting with "C" and negative "Quantity" values
retail_transactions[~(retail_transactions["Invoice"].str.startswith("C", na=False)) & 
                    (retail_transactions["Quantity"] < 0)].reset_index(drop=True)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
1,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
2,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
3,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
4,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
3452,581210,23395,check,-26,2011-12-07 18:36:00,0.0,NaN,United Kingdom
3453,581212,22578,lost,-1050,2011-12-07 18:38:00,0.0,NaN,United Kingdom
3454,581213,22576,check,-30,2011-12-07 18:38:00,0.0,NaN,United Kingdom
3455,581226,23090,missing,-338,2011-12-08 09:56:00,0.0,NaN,United Kingdom


In [ ]:
# Identifying rows with "Invoice" not starting with "C" and negative "Quantity" values and null "Customer ID"
retail_transactions[~(retail_transactions["Invoice"].str.startswith("C", na=False)) & 
                    (retail_transactions["Quantity"] < 0) & 
                    (retail_transactions["Customer ID"].isnull())].reset_index(drop=True)

,Invoice,StockCode,Description,Quantity,InvoiceDate,Price,Customer ID,Country
0,489464,21733,85123a mixed,-96,2009-12-01 10:52:00,0.0,NaN,United Kingdom
1,489463,71477,short,-240,2009-12-01 10:52:00,0.0,NaN,United Kingdom
2,489467,85123A,21733 mixed,-192,2009-12-01 10:53:00,0.0,NaN,United Kingdom
3,489521,21646,NaN,-50,2009-12-01 11:44:00,0.0,NaN,United Kingdom
4,489655,20683,NaN,-44,2009-12-01 17:26:00,0.0,NaN,United Kingdom
...,...,...,...,...,...,...,...,...
3452,581210,23395,check,-26,2011-12-07 18:36:00,0.0,NaN,United Kingdom
3453,581212,22578,lost,-1050,2011-12-07 18:38:00,0.0,NaN,United Kingdom
3454,581213,22576,check,-30,2011-12-07 18:38:00,0.0,NaN,United Kingdom
3455,581226,23090,missing,-338,2011-12-08 09:56:00,0.0,NaN,United Kingdom


In [ ]:
# Inspecting date ranges for each DataFrame
print("table: year_2009_2010")
print(f"start date : {df1['InvoiceDate'].min()}")
print(f"end date : {df1['InvoiceDate'].max()}")

print("-" * 40)

print("table: year_2010_2011")
print(f"start date : {df2['InvoiceDate'].min()}")
print(f"end date : {df2['InvoiceDate'].max()}")

table: year_2009_2010
start date : 2009-12-01 07:45:00
end date : 2010-12-09 20:01:00
----------------------------------------
table: year_2010_2011
start date : 2010-12-01 08:26:00
end date : 2011-12-09 12:50:00


For the year 2009-2010 the date ranges from 01-12-2009 to 09-12-2010

For the year 2009-2010 the date ranges from 01-12-2010 to 09-12-2011

In [ ]:
# Calculating the total number of unique customers in the retail_transactions DataFrame
print(f"total customers : {retail_transactions['Customer ID'].nunique()}")

total customers : 5942


In [ ]:
# Calculating the total number of unique customers in each year
print(f"total customers for the year 2009-2010: {df1['Customer ID'].nunique()}")

print(f"total customers for the year 2010-2011: {df2['Customer ID'].nunique()}")

total customers for the year 2009-2010: 4383
total customers for the year 2010-2011: 4372


In [ ]:
# Calculating the total number of unique countries in the retail_transactions DataFrame and printing the unique country names
print(f" total countries from which customers belong to: {retail_transactions['Country'].nunique()}")

print(f"countries name: {retail_transactions['Country'].unique()}")

 total countries from which customers belong to: 43
countries name: <StringArray>
[      'United Kingdom',               'France',                  'USA',              'Belgium',
            'Australia',                 'EIRE',              'Germany',             'Portugal',
                'Japan',              'Denmark',              'Nigeria',          'Netherlands',
               'Poland',                'Spain',      'Channel Islands',                'Italy',
               'Cyprus',               'Greece',               'Norway',              'Austria',
               'Sweden', 'United Arab Emirates',              'Finland',          'Switzerland',
          'Unspecified',                'Malta',              'Bahrain',                  'RSA',
              'Bermuda',            'Hong Kong',            'Singapore',             'Thailand',
               'Israel',            'Lithuania',          'West Indies',              'Lebanon',
                'Korea',               'Brazi

In [ ]:
# Calculating the total number of unique countries in each year and printing the unique country names
print(f"countries from which people shopped in the year 2009-2010: {df1['Country'].nunique()}")
print(f"countries name: {df1['Country'].unique()}")

print("-" * 120)

print(f"countries from which people shopped in the year 2010-2011: {df2['Country'].nunique()}")
print(f"countries name: {df2['Country'].unique()}")

countries from which people shopped in the year 2009-2010: 40
countries name: <StringArray>
[      'United Kingdom',               'France',                  'USA',              'Belgium',
            'Australia',                 'EIRE',              'Germany',             'Portugal',
                'Japan',              'Denmark',              'Nigeria',          'Netherlands',
               'Poland',                'Spain',      'Channel Islands',                'Italy',
               'Cyprus',               'Greece',               'Norway',              'Austria',
               'Sweden', 'United Arab Emirates',              'Finland',          'Switzerland',
          'Unspecified',                'Malta',              'Bahrain',                  'RSA',
              'Bermuda',            'Hong Kong',            'Singapore',             'Thailand',
               'Israel',            'Lithuania',          'West Indies',              'Lebanon',
                'Korea',           

In [ ]:
# Calculating the total number of distinct products sold in the retail_transactions DataFrame
print(f"distinct products sold : {retail_transactions['Description'].nunique()}")

distinct products sold : 5698


In [ ]:
# Calculating the total number of distinct products sold in each year
print(f"distinct products sold in the year 2009-2010: {df1['Description'].nunique()}")

print(f"distinct products sold in the year 2010-2011: {df2['Description'].nunique()}")

distinct products sold in the year 2009-2010: 4681
distinct products sold in the year 2010-2011: 4223


We have a total of 5942 customers shopping 5698 distinct products from us across 43 different countries in both the years (2009-2010, 2010-2011)

From the above profiling the number says few members droped in the 2nd year